pkg/issuer/acme/acme.go

```diff
@@ -185,23 +186,40 @@ func (a *Acme) createOrder(ctx context.Context, cl client.Interface, crt *v1alph
 		return nil, err
 	}
 
-	// The ACME protocol supports custom validity dates in the certificate order.
-	// pebble and boulder do not support them at this time.
 	if a.issuer.GetSpec().Duration.Duration != 0 {
 		now := time.Now().UTC()
 		order.NotBefore = now
 		order.NotAfter = now.Add(a.issuer.GetSpec().Duration.Duration)
 	}
 
-	order, err = cl.CreateOrder(ctx, order)
+	createdOrder, err := cl.CreateOrder(ctx, order)
 	if err != nil {
+		if ae, ok := err.(*acme.Error); ok &&
+			ae.Type == "urn:ietf:params:acme:error:malformed" &&
+			a.issuer.GetSpec().Duration.Duration != 0 &&
+			strings.Contains(strings.ToLower(ae.Detail), "notbefore") {
+			// The ACME protocol supports user specified validity dates in the certificate order.
+			// Pebble and Boulder do not support them at this time.
+			// ACMEv2 returns a validation error if the server implementation does not support NotBefore and NotAfter.
+			glog.Infof("Error creating order with user specified duration. Retrying order without unsupported parameters: %v", err)
+			a.recorder.Eventf(crt, corev1.EventTypeWarning, "ErrCreateOrder", "Error creating order with user specified duration. Retrying order without unsupported parameters: %v", err)
+
+			// Clearing the specified validity period in the request and try again
+			order.NotBefore = time.Time{}
+			order.NotAfter = time.Time{}
+			createdOrder, err = cl.CreateOrder(ctx, order)
```

Comment:
I think we need some way to mark a certification configuration as invalid in order to prevent re-processing of a Certificate in this case, instead of re-attempting to create ourselves.

This will require us somehow hashing the 'input values' (i.e. the Duration field right now) and recording on the Certificate resource that the particular configuration is invalid. When a re-sync happens (i.e. when the resource changes), we will have to compare that hash to the hash of the new resource in order to determine if a user has updated their configuration, in which case we retry CreateOrder with the new configuration.

I think there are a number of places where this could be useful, at least for ACME. Perhaps for other Issuers (or other parts of the ACME issuer) in future too. It's sort of 'runtime validation' against the actual Issuer, and not just validating the resources are valid 😄

Reply:
Yes I agree. Perhaps we can store it in a label like the hash computed by the deployment resource? In the pod resource this field is called *pod-template-hash*.

Reply:
#761 will allow us to skip processing Certificate resources with particular fields set if they reference ACME issuers.

This should allow us to just blanket not support setting these fields with all ACME issuers. We can then explore at a later time an implementation that's more complex without making a breaking change 😄

Reply:
#761 has now merged, so we should be able to add in this check now 😄 

@vdesjardins are you still in a position to rebase and get this PR up to date? There's been a few changes that may effect your implementation slightly!

Reply:
I should have some time hopefully to start looking at it the coming days!